link-verifier/verify-links.py

```diff
@@ -0,0 +1,396 @@
+        # Only print the file name once
+        file_printed = False
+        for link in self.internal_links:
+            # First, look for anchors in the same document.
+            link_elements = link.split('#')
+            path = link_elements[0]
+            id = None
+            if len(link_elements) > 1:
+                id = link_elements[1]
+            if path == '':
+                if id is not None:
+                    if id.lower() not in self.ids:
+                        self.broken_links.append(link)
+                        file_printed = self.print_filename(files[self.name], file_printed)
+                        cprint(f'\tUnknown link: {link}', 'red')
+                    elif verbose:
+                        file_printed = self.print_filename(files[self.name], file_printed)
+                        cprint(f'\t{link}', 'green')
+                continue
+
+            # At this point, this is probably a link to a file in the same repo,
+            # so we test if the file exists.
+            filename = os.path.join(dirname, path)
+            absfile = os.path.abspath(filename)
+            # Note: We don't test whether the link target exists, just the file.
+            if not os.path.exists(absfile):
+                self.broken_links.append(link)
+                file_printed = self.print_filename(files[self.name], file_printed)
+                cprint(f'\tUnknown file: {path}', 'red')
+            elif verbose:
+                file_printed = self.print_filename(files[self.name], file_printed)
+                cprint(f'\t{link}','green')
+
+        for link in self.external_links:
+            is_broken, status_code = test_url(link)
+            if is_broken:
+                self.broken_links.append(link)
+                file_printed = self.print_filename(files[self.name], file_printed)
+                cprint(f'  {status_code}\t{link}', 'red')
+            else:
+                if verbose:
+                    file_printed = self.print_filename(files[self.name], file_printed)
+                    cprint(f'  {status_code}\t{link}', 'green')
+
+def parse_file(html_file):
+    """Parse href tags from an HTML file"""
+    return HtmlFile(html_file)
+
+def html_name_from_markdown(filename):
+    md_pattern = re.compile('.md', re.IGNORECASE)
+    return md_pattern.sub('.html', filename)
+
+def create_html(markdown_file):
+    """Use pandoc to convert a markdown file to an HTML file"""
+    html_file = html_name_from_markdown(markdown_file)
+    # Convert from Github-flavored Markdown to HTML
+    cmd = f'pandoc -f gfm -o {html_file} {markdown_file}'
+    # Use pandoc to generate HTML from Markdown
+    process = subprocess.run(
+        cmd,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.STDOUT,
+        shell=True,
+        encoding="utf-8",
+        universal_newlines=True
+    )
+    return process
+
+def test_url(url):
+    """Tests a single url"""
+
+    global use_gh_cache
```

Comment:
Do these really need to be global? I think a lot of this could be simplifed with `functools.cache`